enterprise/coderd/usage/publisher.go

```diff
@@ -0,0 +1,427 @@
+		accepted, err := p.publishOnce(publishCtx, deploymentID)
+		publishCtxCancel()
+		if err != nil {
+			return xerrors.Errorf("publish usage events to tallyman: %w", err)
+		}
+		if accepted < tallymanPublishBatchSize {
+			// We published less than the batch size, so we're done.
+			return nil
+		}
+	}
+}
+
+// publishOnce publishes up to tallymanPublishBatchSize usage events to
+// tallyman. It returns the number of successfully published events.
+func (p *tallymanPublisher) publishOnce(ctx context.Context, deploymentID uuid.UUID) (int, error) {
+	licenseJwt, err := p.getBestLicenseJWT(ctx)
+	if xerrors.Is(err, errUsagePublishingDisabled) {
+		return 0, nil
+	} else if err != nil {
+		return 0, xerrors.Errorf("find usage publishing license: %w", err)
+	}
+
+	events, err := p.db.SelectUsageEventsForPublishing(ctx, dbtime.Time(p.clock.Now()))
+	if err != nil {
+		return 0, xerrors.Errorf("select usage events for publishing: %w", err)
+	}
+	if len(events) == 0 {
+		// No events to publish.
+		return 0, nil
+	}
+
+	var (
+		eventIDs    = make(map[string]struct{})
+		tallymanReq = TallymanIngestRequestV1{
+			DeploymentID: deploymentID,
+			Events:       make([]TallymanIngestEventV1, 0, len(events)),
+		}
+	)
+	for _, event := range events {
+		eventIDs[event.ID] = struct{}{}
+		tallymanReq.Events = append(tallymanReq.Events, TallymanIngestEventV1{
+			ID:        event.ID,
+			EventType: event.EventType,
+			EventData: event.EventData,
+			CreatedAt: event.CreatedAt,
+		})
+	}
+	if len(eventIDs) != len(events) {
+		// This should never happen due to the unique constraint in the
+		// database.
+		return 0, xerrors.Errorf("duplicate event IDs found in events for publishing")
+	}
+
+	resp, err := p.sendPublishRequest(ctx, licenseJwt, tallymanReq)
+	allFailed := err != nil
+	if err != nil {
+		p.log.Warn(ctx, "failed to send publish request to tallyman", slog.F("count", len(events)), slog.Error(err))
+		// Fake a response with all events temporarily rejected.
+		resp = TallymanIngestResponseV1{
+			AcceptedEvents: []TallymanIngestAcceptedEventV1{},
+			RejectedEvents: make([]TallymanIngestRejectedEventV1, len(events)),
+		}
+		for i, event := range events {
+			resp.RejectedEvents[i] = TallymanIngestRejectedEventV1{
+				ID:        event.ID,
+				Message:   fmt.Sprintf("failed to publish to tallyman: %v", err),
+				Permanent: false,
+			}
+		}
+	} else {
+		p.log.Debug(ctx, "published usage events to tallyman", slog.F("accepted", len(resp.AcceptedEvents)), slog.F("rejected", len(resp.RejectedEvents)))
+	}
+
+	if len(resp.AcceptedEvents)+len(resp.RejectedEvents) != len(events) {
+		p.log.Warn(ctx, "tallyman returned a different number of events than we sent", slog.F("sent", len(events)), slog.F("accepted", len(resp.AcceptedEvents)), slog.F("rejected", len(resp.RejectedEvents)))
+	}
+
+	var (
+		acceptedEvents = make(map[string]*TallymanIngestAcceptedEventV1)
+		rejectedEvents = make(map[string]*TallymanIngestRejectedEventV1)
+	)
```

Comment:
nit: the multiline `var` makes this harder to read IMO and takes up twice the number of lines as

```
acceptedEvents := make(map[string]*TallymanIngestAcceptedEventV1)
rejectedEvents := make(map[string]*TallymanIngestRejectedEventV1)
```